File: 351002/Yarakhovich/lib/storage/src/postgres/label.rs
use std::sync::Arc;

use domain::entities::{IDType, label::Label};
use sqlx::{Error, PgPool};

#[derive(Clone)]
pub struct PgLabelRepo {
    store: Arc<PgPool>,
}

impl PgLabelRepo {
    pub fn new(store: Arc<PgPool>) -> Self {
        Self { store }
    }
}

impl PgLabelRepo {
    pub async fn create(&self, entity: Label) -> Result<Label, Error> {
        sqlx::query_as::<_, Label>(
            r#"INSERT INTO tbl_label (name)
               VALUES ($1)
               RETURNING id, name"#,
        )
        .bind(entity.name)
        .fetch_one(&*self.store)
        .await
    }

    pub async fn get(&self, id: IDType) -> Result<Option<Label>, Error> {
        sqlx::query_as::<_, Label>("SELECT id, name FROM tbl_label WHERE id = $1")
            .bind(id)
            .fetch_optional(&*self.store)
            .await
    }

    pub async fn update(&self, id: IDType, entity: Label) -> Result<Label, Error> {
        sqlx::query_as::<_, Label>(
            r#"UPDATE tbl_label
               SET name = $2
               WHERE id = $1
               RETURNING id, name"#,
        )
        .bind(id)
        .bind(entity.name)
        .fetch_one(&*self.store)
        .await
    }

    pub async fn delete(&self, id: IDType) -> Result<(), Error> {
        let result = sqlx::query("DELETE FROM tbl_label WHERE id = $1")
            .bind(id)
            .execute(&*self.store)
            .await?;
        if result.rows_affected() == 0 {
            return Err(Error::RowNotFound);
        }
        Ok(())
    }

    pub async fn list(&self) -> Result<Vec<Label>, Error> {
        sqlx::query_as::<_, Label>("SELECT id, name FROM tbl_label")
            .fetch_all(&*self.store)
            .await
    }
}

File: 351001/Ushakov/lab2/internal/storage/label/label.go
package label

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"github.com/Khmelov/Distcomp/351001/Ushakov/lab2/internal/model"
	"github.com/jmoiron/sqlx"
	"github.com/lib/pq"
)

var (
	Errgithub.com/Khmelov/Distcomp/351001/UshakovNotFound = fmt.Errorf(
"post not found"
)
ErrConstraintsCheck = fmt.Errorf("invalid data passed")
)
type repo struct {
	db *sqlx.DB
}

type Repo interface {
	Create(ctx context.Context, req model.Label) (model.Label, error)
	GetList(ctx context.Context) ([]model.Label, error)
	Get(ctx context.Context, id int64) (model.Label, error)
	Update(ctx context.Context, req model.Label) (model.Label, error)
	Delete(ctx context.Context, id int64) error
}

func New(db *sqlx.DB) Repo {
	return repo{
		db: db,
	}
}

func (r repo) Create(ctx context.Context, req model.Label) (model.Label, error) {
	query := `INSERT INTO tbl_label (name) 
	          VALUES ($1) RETURNING id`

	var id int64

	err := r.db.QueryRowContext(ctx, query, req.Name).Scan(&id)
	if err != nil {
		if pqErr, ok := err.(*pq.Error); ok && pqErr.Code == "23505" {
			return model.Label{}, ErrConstraintsCheck
		}

		return model.Label{}, fmt.Errorf("failed to create Repo: %w", err)
	}

	req.ID = id

	return req, nil
}

func (r repo) Delete(ctx context.Context, id int64) error {
	query := `DELETE FROM tbl_label WHERE id = $1`

	result, err := r.db.ExecContext(ctx, query, id)
	if err != nil {
		return fmt.Errorf("failed to delete Repo: %w", err)
	}

	rowsAffected, err := result.RowsAffected()
	if err != nil {
		return fmt.Errorf("failed to check rows affected: %w", err)
	}

	if rowsAffected == 0 {
		return Errgithub.com / Khmelov / Distcomp / 351001 / UshakovNotFound
	}

	return nil
}

func (r repo) Get(ctx context.Context, id int64) (model.Label, error) {
	var result model.Label

	query := `SELECT * FROM tbl_label WHERE id = $1`

	err := r.db.GetContext(ctx, &result, query, id)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return result, Errgithub.com / Khmelov / Distcomp / 351001 / UshakovNotFound
		}
		return result, fmt.Errorf("failed to retrieve Repo by ID: %w", err)
	}

	return result, nil
}

func (r repo) GetList(ctx context.Context) ([]model.Label, error) {
	var result []model.Label

	query := `SELECT * FROM tbl_label`

	err := r.db.SelectContext(ctx, &result, query)
	if err != nil {
		return nil, fmt.Errorf("failed to retrieve result: %w", err)
	}

	if len(result) == 0 {
		return []model.Label{}, nil
	}

	return result, nil
}

func (r repo) Update(ctx context.Context, req model.Label) (model.Label, error) {
	query := `UPDATE tbl_label SET name = $1
	          WHERE id = $2 RETURNING id, name`

	var result model.Label

	err := r.db.QueryRowContext(ctx, query, req.Name, req.ID).Scan(&result.ID, &result.Name)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return result, Errgithub.com / Khmelov / Distcomp / 351001 / UshakovNotFound
		}

		return result, fmt.Errorf("failed to update Repo: %w", err)
	}

	return result, nil
}
